### ISSUE TITLE
UIO Driver fails to recognize that it manages NVMe disks

### ISSUE THREAD

**Issue by patr-geary-smci** [spdk/spdk#273]
Testing with 94de7e91755a2d6b3b4a6a94003275f0966467b0

These are SAMSUNG MZ4LB3T8HALS-00003 M.3 NVMe SSD's; I sadly have nothing else on hand at the moment.  When I am able to get 32x M.3 Intel disks I'll give it another go and see if things behave.

Anyways, to the problem,

With UIO, some of the disks don't init correctly when executing perf:
EAL: PCI device 0000:44:00.0 on NUMA socket 0
EAL:   probe driver: 144d:a808 spdk_nvme
EAL:   0000:44:00.0 not managed by UIO driver, skipping
EAL: Driver cannot attach the device (0000:44:00.0)

Setup.sh shows that it's allocating the aforementioned device and reports no error:
[root@105-91 spdk]# NRHUGE=128 ./scripts/setup.sh
0000:1a:00.0 (144d a808): nvme -> uio_pci_generic
...
0000:44:00.0 (144d a808): nvme -> uio_pci_generic
...

Single disk attempts fail as well; it proclaims that it does not controlled by UIO Driver in exactly the same way.

It seems semi-arbitrary; some of the NVMe's on the pcie expanders will initialize correctly, others will not.

With intel_iommu=on using vfio; it's 100% OK for all disks:
EAL: PCI device 0000:44:00.0 on NUMA socket 0
EAL:   probe driver: 144d:a808 spdk_nvme
Attaching to NVMe Controller at 0000:44:00.0 [144d:a808]
Attached to NVMe Controller at 0000:44:00.0 [144d:a808]

Setup sequence:
[root@105-91 spdk]# NRHUGE=128 ./scripts/setup.sh
0000:1a:00.0 (144d a808): nvme -> vfio-pci
...
0000:44:00.0 (144d a808): nvme -> vfio-pci
...

See attached archive for lspci dumps; complete messages from perf/setup against all disks.  Should there be anything else of importance; ping me.

[vfio_uio_discrepancies_2018-03-30.tar.gz](https://github.com/spdk/spdk/files/1864517/vfio_uio_discrepancies_2018-03-30.tar.gz)

**Comment by jimharris** [comment]
What do you see for each of the following:

ls –lL /sys/bus/pci/devices/0000:20:00.0
ls –lL /sys/bus/pci/devices/0000:21:00.0

(that’s a lower case L and an upper case L)

-Jim

**Comment by patr-geary-smci** [comment]
[root@105-91 spdk]# ls -lL /sys/bus/pci/devices/0000\:21\:00.0/
total 0
-rw-r--r--. 1 root root  4096 Mar 30 12:38 broken_parity_status
-r--r--r--. 1 root root  4096 Mar 30 11:10 class
-rw-r--r--. 1 root root  4096 Mar 30 11:10 config
-r--r--r--. 1 root root  4096 Mar 30 12:38 consistent_dma_mask_bits
-rw-r--r--. 1 root root  4096 Mar 30 12:38 d3cold_allowed
-r--r--r--. 1 root root  4096 Mar 30 11:10 device
-r--r--r--. 1 root root  4096 Mar 30 12:38 dma_mask_bits
drwxr-xr-x. 2 root root     0 Mar 30 11:21 driver
-rw-r--r--. 1 root root  4096 Mar 30 12:38 driver_override
-rw-r--r--. 1 root root  4096 Mar 30 12:38 enable
-r--r--r--. 1 root root  4096 Mar 30 11:10 irq
-r--r--r--. 1 root root  4096 Mar 30 12:38 local_cpulist
-r--r--r--. 1 root root  4096 Mar 30 11:10 local_cpus
-r--r--r--. 1 root root  4096 Mar 30 11:26 modalias
-rw-r--r--. 1 root root  4096 Mar 30 12:38 msi_bus
-rw-r--r--. 1 root root  4096 Mar 30 11:10 numa_node
drwxr-xr-x. 2 root root     0 Mar 30 12:38 power
--w--w----. 1 root root  4096 Mar 30 12:38 remove
--w--w----. 1 root root  4096 Mar 30 12:38 rescan
--w-------. 1 root root  4096 Mar 30 12:38 reset
-r--r--r--. 1 root root  4096 Mar 30 11:10 resource
-rw-------. 1 root root 16384 Mar 30 12:38 resource0
drwxr-xr-x. 5 root root     0 Mar 30 11:09 subsystem
-r--r--r--. 1 root root  4096 Mar 30 11:27 subsystem_device
-r--r--r--. 1 root root  4096 Mar 30 11:27 subsystem_vendor
-rw-r--r--. 1 root root  4096 Mar 30 11:09 uevent
-r--r--r--. 1 root root  4096 Mar 30 11:10 vendor
[root@105-91 spdk]# ls -lL /sys/bus/pci/devices/0000\:20\:00.0/
total 0
-rw-r--r--. 1 root root  4096 Mar 30 12:36 broken_parity_status
-r--r--r--. 1 root root  4096 Mar 30 11:10 class
-rw-r--r--. 1 root root  4096 Mar 30 11:10 config
-r--r--r--. 1 root root  4096 Mar 30 12:36 consistent_dma_mask_bits
-rw-r--r--. 1 root root  4096 Mar 30 12:36 d3cold_allowed
-r--r--r--. 1 root root  4096 Mar 30 11:10 device
-r--r--r--. 1 root root  4096 Mar 30 12:36 dma_mask_bits
drwxr-xr-x. 2 root root     0 Mar 30 11:21 driver
-rw-r--r--. 1 root root  4096 Mar 30 12:36 driver_override
-rw-r--r--. 1 root root  4096 Mar 30 12:36 enable
-r--r--r--. 1 root root  4096 Mar 30 11:10 irq
-r--r--r--. 1 root root  4096 Mar 30 12:36 local_cpulist
-r--r--r--. 1 root root  4096 Mar 30 11:10 local_cpus
-r--r--r--. 1 root root  4096 Mar 30 11:26 modalias
-rw-r--r--. 1 root root  4096 Mar 30 12:36 msi_bus
-rw-r--r--. 1 root root  4096 Mar 30 11:10 numa_node
drwxr-xr-x. 2 root root     0 Mar 30 12:36 power
--w--w----. 1 root root  4096 Mar 30 12:36 remove
--w--w----. 1 root root  4096 Mar 30 12:36 rescan
--w-------. 1 root root  4096 Mar 30 12:36 reset
-r--r--r--. 1 root root  4096 Mar 30 11:10 resource
-rw-------. 1 root root 16384 Mar 30 11:27 resource0
drwxr-xr-x. 5 root root     0 Mar 30 11:09 subsystem
-r--r--r--. 1 root root  4096 Mar 30 11:27 subsystem_device
-r--r--r--. 1 root root  4096 Mar 30 11:27 subsystem_vendor
-rw-r--r--. 1 root root  4096 Mar 30 11:09 uevent
drwxr-xr-x. 3 root root     0 Mar 30 12:37 uio
-r--r--r--. 1 root root  4096 Mar 30 11:10 vendor

Yeah; looks like the uio sysfs entry isn't popping up for 21.

**Comment by wanqunintel** [comment]
It seems we don't have the M.3 ssd here to verify it, does anyone has this environment?

**Comment by wanqunintel** [comment]
This should not related to the SPDK issue per the team meeting discussion

**Comment by patr-geary-smci** [comment]
I've actually been waiting on system availability so that I can slap it onto something you can access.


Sadly the M.3's have been subjected to repeated requests for performance data so we haven't had a chance.


Lemme ping them again and see if I can get access for a 48 hour time frame.  You'll need to confer with me through masked channels (email); I have little desire to make remote system credentials public.


Thank you,

Patrick Geary
Software Engineer
Cell: (408) 390-7124
Skype: patr.geary.sm
________________________________

**Comment by jimharris** [comment]
Closing this issue as a duplicate of #477.  #477 has many more details around analysis.
